evalml/tests/automl_tests/test_time_series_split.py

```diff
@@ -0,0 +1,47 @@
+import pandas as pd
+import pytest
+
+from evalml.automl.data_splitters import TimeSeriesSplit
+
+
+def test_time_series_split_init():
+    ts_split = TimeSeriesSplit(gap=3, max_delay=4, n_folds=5)
+    assert ts_split.get_n_splits() == 5
+
+    with pytest.raises(ValueError, match="Both X and y cannot be None or empty in TimeSeriesSplit.split"):
+        _ = list(ts_split.split(X=None, y=None))
+
+    with pytest.raises(ValueError, match="Both X and y cannot be None or empty in TimeSeriesSplit.split"):
+        _ = list(ts_split.split(X=pd.DataFrame(), y=pd.Series([])))
+
+
+@pytest.mark.parametrize("max_delay,gap", [(0, 0), (1, 0), (2, 0), (0, 3), (1, 1), (4, 2)])
+@pytest.mark.parametrize("X_none,y_none", [(False, False), (True, False), (False, True)])
+def test_time_series_split(max_delay, gap, X_none, y_none):
+    X = pd.DataFrame({"features": range(1, 32)})
+    y = pd.Series(range(1, 32))
+
+    # Splitter does not need a daterange index. We use a daterange index so that the
+    # expected answer is easier to understand
+    y.index = pd.date_range("2020-10-01", "2020-10-31")
+    X.index = pd.date_range("2020-10-01", "2020-10-31")
+
+    answer = [(pd.date_range("2020-10-01", f"2020-10-{10 + gap}"), pd.date_range(f"2020-10-{11 - max_delay}", f"2020-10-{17 + gap}")),
+              (pd.date_range("2020-10-01", f"2020-10-{17 + gap}"), pd.date_range(f"2020-10-{18 - max_delay}", f"2020-10-{24 + gap}")),
+              (pd.date_range("2020-10-01", f"2020-10-{24 + gap}"), pd.date_range(f"2020-10-{25 - max_delay}", "2020-10-31"))]
```

Comment:
Ah neat, I didn't expect the answers here to be this cleanly defined.

Considering the case when `max_delay=0` and `gap=0` for simplicity, why does the first training split range from 10-01 to 10-10 excluding the last day (i.e. 9 days), but the test split always has 6 days, and each subsequent update increases by 7 days? I guess my question is, why aren't all these splits the same size?

I think the simplest algo to follow for this would be if we divide the data into `n_splits` equal sections, and then each split in the generator uses the i-th ordered section as the test data, and the concatenation of all previous ordered sections as the training data. It appears the algo we've currently got matches this in the ordering, but the section sizes are a bit different. How come?

Reply:
I think the reason here is that the number of datapoints (31) is not divisible by the number of folds (4 because splitting three times creates four chunks of data). 

Splitting the data equally into four creates would results in four 7-day chunks of data with three days leftover. We just add the first three days to the first chunk of data. 

I think if the data isn't evenly divisible by the number of splits, you'll always end up in a situation where one of the splits is slightly larger (unless you throw out data which I don't think we should do).

If make a dataframe with 32 rows instead of 31 and use n_splits=3, the first train/test are 8 days long and each subsequent train set increases by 8 days.

Reply:
@freddyaboulton got it, thanks for the explanation. I wanted to make sure there wasn't a bug lurking in there somewhere. I would've guessed the underlying sklearn splits would be as even as possible but I guess not! It'll be interesting to learn how this behaves on other datasets.